File: sdk/javascript/kubesentiment-sdk.ts
/**
 * KubeSentiment TypeScript/JavaScript SDK
 *
 * Official TypeScript/JavaScript client library for KubeSentiment API.
 *
 * @example
 * ```typescript
 * import { KubeSentimentClient } from 'kubesentiment-sdk';
 *
 * const client = new KubeSentimentClient({
 *   baseUrl: 'http://localhost:8000',
 *   apiKey: 'your-api-key'
 * });
 *
 * const result = await client.predict('This is amazing!');
 * console.log(result.label, result.confidence);
 * ```
 */

export const VERSION = '1.0.0';

/**
 * Sentiment prediction labels
 */
export enum PredictionLabel {
  POSITIVE = 'POSITIVE',
  NEGATIVE = 'NEGATIVE',
  NEUTRAL = 'NEUTRAL'
}

/**
 * Batch job priority levels
 */
export enum Priority {
  HIGH = 'high',
  MEDIUM = 'medium',
  LOW = 'low'
}

/**
 * Batch job status
 */
export enum JobStatus {
  PENDING = 'pending',
  PROCESSING = 'processing',
  COMPLETED = 'completed',
  FAILED = 'failed'
}

/**
 * Single prediction result
 */
export interface PredictionResult {
  label: string;
  confidence: number;
  inferenceTimeMs: number;
  modelName: string;
  backend: string;
  cached: boolean;
  textLength?: number;
  explanation?: Record<string, any>;
}

/**
 * Batch prediction job
 */
export interface BatchJob {
  jobId: string;
  status: string;
  createdAt: string;
  priority: string;
  totalTexts: number;
  processed?: number;
  results?: PredictionResult[];
  error?: string;
}

/**
 * Model information
 */
export interface ModelInfo {
  modelName: string;
  backend: string;
  version?: string;
  loaded: boolean;
}

/**
 * Health check response
 */
export interface HealthStatus {
  status: string;
  timestamp: string;
  version?: string;
}

/**
 * Client configuration options
 */
export interface ClientConfig {
  /** Base URL of the KubeSentiment API */
  baseUrl: string;
  /** Optional API key for authentication */
  apiKey?: string;
  /** Request timeout in milliseconds (default: 30000) */
  timeout?: number;
  /** Custom headers */
  headers?: Record<string, string>;
}

/**
 * Prediction options
 */
export interface PredictOptions {
  /** Whether to include model explanation */
  explain?: boolean;
  /** Request timeout in milliseconds */
  timeout?: number;
}

/**
 * Batch prediction options
 */
export interface BatchPredictOptions {
  /** Job priority */
  priority?: Priority;
  /** Whether to wait for job completion */
  wait?: boolean;
  /** Polling interval in milliseconds (if wait=true) */
  pollInterval?: number;
}

/**
 * Custom error class for SDK errors
 */
export class KubeSentimentError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'KubeSentimentError';
  }
}

/**
 * API error class
 */
export class APIError extends KubeSentimentError {
  statusCode: number;

  constructor(statusCode: number, message: string) {
    super(`API Error ${statusCode}: ${message}`);
    this.statusCode = statusCode;
    this.name = 'APIError';
  }
}

/**
 * Official TypeScript/JavaScript client for KubeSentiment API
 *
 * @example
 * ```typescript
 * const client = new KubeSentimentClient({
 *   baseUrl: 'http://localhost:8000',
 *   apiKey: 'your-api-key'
 * });
 *
 * // Single prediction
 * const result = await client.predict('I love this product!');
 * console.log(`${result.label}: ${(result.confidence * 100).toFixed(2)}%`);
 *
 * // Batch prediction
 * const job = await client.batchPredict(['Great!', 'Terrible'], { wait: true });
 * job.results?.forEach(r => console.log(r.label));
 * ```
 */
export class KubeSentimentClient {
  private baseUrl: string;
  private apiKey?: string;
  private timeout: number;
  private headers: Record<string, string>;

  constructor(config: ClientConfig) {
    this.baseUrl = config.baseUrl.replace(/\/$/, '');
    this.apiKey = config.apiKey;
    this.timeout = config.timeout || 30000;
    this.headers = {
      'Content-Type': 'application/json',
      'User-Agent': `kubesentiment-sdk/${VERSION}`,
      ...config.headers
    };

    if (this.apiKey) {
      this.headers['X-API-Key'] = this.apiKey;
    }
  }

  /**
   * Make HTTP request to API
   */
  private async request<T>(
    method: string,
    endpoint: string,
    body?: any,
    options?: { timeout?: number }
  ): Promise<T> {
    const url = `${this.baseUrl}${endpoint}`;
    const timeout = options?.timeout || this.timeout;

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    try {
      const response = await fetch(url, {
        method,
        headers: this.headers,
        body: body ? JSON.stringify(body) : undefined,
        signal: controller.signal
      });

      clearTimeout(timeoutId);

      if (!response.ok) {
        let errorMessage: string;
        try {
          const errorData = await response.json();
          errorMessage = errorData.detail || response.statusText;
        } catch {
          errorMessage = response.statusText;
        }
        throw new APIError(response.status, errorMessage);
      }

      return await response.json();
    } catch (error) {
      clearTimeout(timeoutId);
      if (error instanceof APIError) {
        throw error;
      }
      if (error instanceof Error) {
        if (error.name === 'AbortError') {
          throw new KubeSentimentError(`Request timeout after ${timeout}ms`);
        }
        throw new KubeSentimentError(`Request failed: ${error.message}`);
      }
      throw new KubeSentimentError('Request failed with unknown error');
    }
  }

  /**
   * Predict sentiment for a single text
   *
   * @param text - Input text to analyze
   * @param options - Prediction options
   * @returns Promise resolving to PredictionResult
   *
   * @example
   * ```typescript
   * const result = await client.predict('This is great!');
   * console.log(result.label, result.confidence);
   * ```
   */
  async predict(text: string, options?: PredictOptions): Promise<PredictionResult> {
    const response = await this.request<any>(
      'POST',
      '/api/v1/predict',
      { text, explain: options?.explain || false },
      { timeout: options?.timeout }
    );

    return {
      label: response.label,
      confidence: response.score,
      inferenceTimeMs: response.inference_time_ms || 0,
      modelName: response.model_name || 'unknown',
      backend: response.backend || 'unknown',
      cached: response.cached || false,
      textLength: response.text_length,
      explanation: response.explanation
    };
  }

  /**
   * Submit batch prediction job
   *
   * @param texts - Array of texts to analyze
   * @param options - Batch prediction options
   * @returns Promise resolving to BatchJob
   *
   * @example
   * ```typescript
   * const job = await client.batchPredict(['Great!', 'Terrible'], { wait: true });
   * job.results?.forEach(r => console.log(r.label));
   * ```
   */
  async batchPredict(
    texts: string[],
    options?: BatchPredictOptions
  ): Promise<BatchJob> {
    const priority = options?.priority || Priority.MEDIUM;

    const response = await this.request<any>(
      'POST',
      '/api/v1/batch/predict',
      { texts, priority }
    );

    const job: BatchJob = {
      jobId: response.job_id,
      status: response.status,
      createdAt: response.created_at,
      priority: response.priority,
      totalTexts: response.total_texts
    };

    if (options?.wait) {
      return this.waitForJob(job.jobId, options.pollInterval);
    }

    return job;
  }

  /**
   * Get status of batch job
   *
   * @param jobId - Batch job ID
   * @returns Promise resolving to BatchJob
   */
  async getBatchStatus(jobId: string): Promise<BatchJob> {
    const response = await this.request<any>('GET', `/api/v1/batch/status/${jobId}`);

    return {
      jobId: response.job_id,
      status: response.status,
      createdAt: response.created_at,
      priority: response.priority,
      totalTexts: response.total_texts,
      processed: response.processed
    };
  }

  /**
   * Get results of completed batch job
   *
   * @param jobId - Batch job ID
   * @returns Promise resolving to BatchJob with results
   */
  async getBatchResults(jobId: string): Promise<BatchJob> {
    const response = await this.request<any>('GET', `/api/v1/batch/results/${jobId}`);

    const results: PredictionResult[] = response.results?.map((r: any) => ({
      label: r.label,
      confidence: r.score,
      inferenceTimeMs: r.inference_time_ms || 0,
      modelName: r.model_name || 'unknown',
      backend: r.backend || 'unknown',
      cached: r.cached || false
    })) || [];

    return {
      jobId: response.job_id,
      status: response.status,
      createdAt: response.created_at,
      priority: response.priority,
      totalTexts: response.total_texts,
      results
    };
  }

  /**
   * Wait for batch job to complete
   *
   * @param jobId - Batch job ID
   * @param pollInterval - Polling interval in milliseconds (default: 2000)
   * @param maxWait - Maximum wait time in milliseconds (default: 300000)
   * @returns Promise resolving to completed BatchJob
   */
  async waitForJob(
    jobId: string,
    pollInterval: number = 2000,
    maxWait: number = 300000
  ): Promise<BatchJob> {
    const startTime = Date.now();

    while (true) {
      const job = await this.getBatchStatus(jobId);

      if (job.status === JobStatus.COMPLETED) {
        return this.getBatchResults(jobId);
      } else if (job.status === JobStatus.FAILED) {
        throw new KubeSentimentError(`Batch job failed: ${job.error}`);
      }

      if (Date.now() - startTime > maxWait) {
        throw new KubeSentimentError(
          `Job did not complete within ${maxWait}ms`
        );
      }

      await new Promise(resolve => setTimeout(resolve, pollInterval));
    }
  }

  /**
   * Cancel a batch job
   *
   * @param jobId - Batch job ID
   * @returns Promise resolving to true if cancelled successfully
   */
  async cancelBatchJob(jobId: string): Promise<boolean> {
    try {
      await this.request('DELETE', `/api/v1/batch/${jobId}`);
      return true;
    } catch (error) {
      return false;
    }
  }

  /**
   * Get information about the loaded model
   *
   * @returns Promise resolving to model information
   */
  async getModelInfo(): Promise<ModelInfo> {
    return this.request<ModelInfo>('GET', '/api/v1/model-info');
  }

  /**
   * Check service health
   *
   * @returns Promise resolving to health status
   */
  async healthCheck(): Promise<HealthStatus> {
    return this.request<HealthStatus>('GET', '/api/v1/health');
  }

  /**
   * Get Prometheus metrics
   *
   * @returns Promise resolving to metrics in Prometheus format
   */
  async getMetrics(): Promise<string> {
    const url = `${this.baseUrl}/api/v1/metrics`;
    const response = await fetch(url, { headers: this.headers });
    if (!response.ok) {
      throw new APIError(response.status, response.statusText);
    }
    return response.text();
  }

  /**
   * Get drift detection summary
   *
   * @returns Promise resolving to drift detection statistics
   */
  async getDriftSummary(): Promise<Record<string, any>> {
    return this.request<Record<string, any>>('GET', '/api/v1/monitoring/drift');
  }

  /**
   * Get business KPIs
   *
   * @returns Promise resolving to business metrics
   */
  async getBusinessKPIs(): Promise<Record<string, any>> {
    return this.request<Record<string, any>>('GET', '/api/v1/monitoring/kpis/business');
  }
}

/**
 * Convenience function for quick predictions
 *
 * @param text - Text to analyze
 * @param baseUrl - API base URL (default: http://localhost:8000)
 * @param apiKey - Optional API key
 * @returns Promise resolving to PredictionResult
 *
 * @example
 * ```typescript
 * import { predict } from 'kubesentiment-sdk';
 * const result = await predict('I love this!');
 * console.log(result.label);
 * ```
 */
export async function predict(
  text: string,
  baseUrl: string = 'http://localhost:8000',
  apiKey?: string
): Promise<PredictionResult> {
  const client = new KubeSentimentClient({ baseUrl, apiKey });
  return client.predict(text);
}

File: sdk/python/kubesentiment_sdk.py
"""
KubeSentiment Python SDK

Official Python client library for KubeSentiment API.

Example usage:
    >>> from kubesentiment_sdk import KubeSentimentClient
    >>> client = KubeSentimentClient(base_url="http://localhost:8000", api_key="your-key")
    >>> result = client.predict("This is amazing!")
    >>> print(result.label, result.confidence)
    POSITIVE 0.9876
"""

import requests
from typing import List, Dict, Optional, Any
from dataclasses import dataclass
from enum import Enum
import time


__version__ = "1.0.0"


class PredictionLabel(str, Enum):
    """Sentiment prediction labels."""
    POSITIVE = "POSITIVE"
    NEGATIVE = "NEGATIVE"
    NEUTRAL = "NEUTRAL"


class Priority(str, Enum):
    """Batch job priority levels."""
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


class JobStatus(str, Enum):
    """Batch job status."""
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


@dataclass
class PredictionResult:
    """Single prediction result."""
    label: str
    confidence: float
    inference_time_ms: float
    model_name: str
    backend: str
    cached: bool
    text_length: Optional[int] = None
    explanation: Optional[Dict[str, Any]] = None

    @property
    def is_positive(self) -> bool:
        """Check if prediction is positive."""
        return self.label == PredictionLabel.POSITIVE.value

    @property
    def is_negative(self) -> bool:
        """Check if prediction is negative."""
        return self.label == PredictionLabel.NEGATIVE.value

    @property
    def is_confident(self, threshold: float = 0.8) -> bool:
        """Check if prediction confidence exceeds threshold."""
        return self.confidence >= threshold


@dataclass
class BatchJob:
    """Batch prediction job."""
    job_id: str
    status: str
    created_at: str
    priority: str
    total_texts: int
    processed: Optional[int] = None
    results: Optional[List[PredictionResult]] = None
    error: Optional[str] = None


class KubeSentimentError(Exception):
    """Base exception for KubeSentiment SDK."""
    pass


class APIError(KubeSentimentError):
    """API request failed."""
    def __init__(self, status_code: int, message: str):
        """Initialize the APIError.

        Args:
            status_code: The HTTP status code returned by the API.
            message: The error message returned by the API.
        """
        self.status_code = status_code
        self.message = message
        super().__init__(f"API Error {status_code}: {message}")


class KubeSentimentClient:
    """
    Official Python client for KubeSentiment API.

    Args:
        base_url: Base URL of the KubeSentiment API (e.g., "http://localhost:8000")
        api_key: Optional API key for authentication
        timeout: Request timeout in seconds (default: 30)

    Example:
        >>> client = KubeSentimentClient(base_url="http://localhost:8000")
        >>> result = client.predict("I love this product!")
        >>> print(f"{result.label}: {result.confidence:.2%}")
    """

    def __init__(
        self,
        base_url: str,
        api_key: Optional[str] = None,
        timeout: int = 30,
        verify_ssl: bool = True
    ):
        """Initializes the KubeSentiment client.

        Args:
            base_url: Base URL of the KubeSentiment API (e.g., "http://localhost:8000").
            api_key: Optional API key for authentication.
            timeout: Request timeout in seconds (default: 30).
            verify_ssl: Whether to verify SSL certificates (default: True).
        """
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key
        self.timeout = timeout
        self.verify_ssl = verify_ssl

        # Setup session
        self.session = requests.Session()
        if api_key:
            self.session.headers['X-API-Key'] = api_key
        self.session.headers['User-Agent'] = f'kubesentiment-sdk/{__version__}'

    def _request(
        self,
        method: str,
        endpoint: str,
        json: Optional[Dict] = None,
        params: Optional[Dict] = None
    ) -> Dict[str, Any]:
        """Make HTTP request to API.

        Args:
            method: HTTP method (GET, POST, etc.).
            endpoint: API endpoint path.
            json: JSON body for the request.
            params: Query parameters.

        Returns:
            The parsed JSON response.

        Raises:
            APIError: If the API returns an error status code.
            KubeSentimentError: If the request fails due to network issues.
        """
        url = f"{self.base_url}{endpoint}"

        try:
            response = self.session.request(
                method=method,
                url=url,
                json=json,
                params=params,
                timeout=self.timeout,
                verify=self.verify_ssl
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.HTTPError as e:
            try:
                error_detail = e.response.json().get('detail', str(e))
            except:
                error_detail = str(e)
            raise APIError(e.response.status_code, error_detail)
        except requests.exceptions.RequestException as e:
            raise KubeSentimentError(f"Request failed: {e}")

    def predict(
        self,
        text: str,
        explain: bool = False,
        timeout: Optional[int] = None
    ) -> PredictionResult:
        """
        Predict sentiment for a single text.

        Args:
            text: Input text to analyze
            explain: Whether to include model explanation
            timeout: Request timeout (overrides default)

        Returns:
            PredictionResult object

        Example:
            >>> result = client.predict("This is great!")
            >>> print(result.label, result.confidence)
        """
        # Use original timeout if not overridden
        original_timeout = self.timeout
        if timeout:
            self.timeout = timeout

        try:
            response = self._request(
                'POST',
                '/api/v1/predict',
                json={'text': text, 'explain': explain}
            )

            return PredictionResult(
                label=response['label'],
                confidence=response['score'],
                inference_time_ms=response.get('inference_time_ms', 0),
                model_name=response.get('model_name', 'unknown'),
                backend=response.get('backend', 'unknown'),
                cached=response.get('cached', False),
                text_length=response.get('text_length'),
                explanation=response.get('explanation')
            )
        finally:
            self.timeout = original_timeout

    def batch_predict(
        self,
        texts: List[str],
        priority: Priority = Priority.MEDIUM,
        wait: bool = False,
        poll_interval: int = 2
    ) -> BatchJob:
        """
        Submit batch prediction job.

        Args:
            texts: List of texts to analyze
            priority: Job priority (high, medium, low)
            wait: Whether to wait for job completion
            poll_interval: Polling interval in seconds (if wait=True)

        Returns:
            BatchJob object

        Example:
            >>> texts = ["Great product!", "Terrible service"]
            >>> job = client.batch_predict(texts, wait=True)
            >>> for result in job.results:
            ...     print(result.label)
        """
        response = self._request(
            'POST',
            '/api/v1/batch/predict',
            json={'texts': texts, 'priority': priority.value}
        )

        job = BatchJob(
            job_id=response['job_id'],
            status=response['status'],
            created_at=response['created_at'],
            priority=response['priority'],
            total_texts=response['total_texts']
        )

        if wait:
            return self.wait_for_job(job.job_id, poll_interval=poll_interval)

        return job

    def get_batch_status(self, job_id: str) -> BatchJob:
        """
        Get status of batch job.

        Args:
            job_id: Batch job ID

        Returns:
            BatchJob object
        """
        response = self._request('GET', f'/api/v1/batch/status/{job_id}')

        return BatchJob(
            job_id=response['job_id'],
            status=response['status'],
            created_at=response['created_at'],
            priority=response['priority'],
            total_texts=response['total_texts'],
            processed=response.get('processed')
        )

    def get_batch_results(self, job_id: str) -> BatchJob:
        """
        Get results of completed batch job.

        Args:
            job_id: Batch job ID

        Returns:
            BatchJob object with results
        """
        response = self._request('GET', f'/api/v1/batch/results/{job_id}')

        results = []
        if 'results' in response:
            for r in response['results']:
                results.append(PredictionResult(
                    label=r['label'],
                    confidence=r['score'],
                    inference_time_ms=r.get('inference_time_ms', 0),
                    model_name=r.get('model_name', 'unknown'),
                    backend=r.get('backend', 'unknown'),
                    cached=r.get('cached', False)
                ))

        return BatchJob(
            job_id=response['job_id'],
            status=response['status'],
            created_at=response['created_at'],
            priority=response['priority'],
            total_texts=response['total_texts'],
            results=results
        )

    def wait_for_job(
        self,
        job_id: str,
        poll_interval: int = 2,
        max_wait: int = 300
    ) -> BatchJob:
        """
        Wait for batch job to complete.

        Args:
            job_id: Batch job ID
            poll_interval: Polling interval in seconds
            max_wait: Maximum wait time in seconds

        Returns:
            Completed BatchJob object

        Raises:
            TimeoutError: If job doesn't complete within max_wait
        """
        start_time = time.time()

        while True:
            job = self.get_batch_status(job_id)

            if job.status == JobStatus.COMPLETED.value:
                return self.get_batch_results(job_id)
            elif job.status == JobStatus.FAILED.value:
                raise KubeSentimentError(f"Batch job failed: {job.error}")

            if time.time() - start_time > max_wait:
                raise TimeoutError(f"Job did not complete within {max_wait} seconds")

            time.sleep(poll_interval)

    def cancel_batch_job(self, job_id: str) -> bool:
        """
        Cancel a batch job.

        Args:
            job_id: Batch job ID

        Returns:
            True if cancelled successfully
        """
        try:
            self._request('DELETE', f'/api/v1/batch/{job_id}')
            return True
        except APIError:
            return False

    def get_model_info(self) -> Dict[str, Any]:
        """
        Get information about the loaded model.

        Returns:
            Dictionary with model information
        """
        return self._request('GET', '/api/v1/model-info')

    def health_check(self) -> Dict[str, Any]:
        """
        Check service health.

        Returns:
            Health status dictionary
        """
        return self._request('GET', '/api/v1/health')

    def get_metrics(self) -> str:
        """
        Get Prometheus metrics.

        Returns:
            Metrics in Prometheus format
        """
        url = f"{self.base_url}/api/v1/metrics"
        response = self.session.get(url, timeout=self.timeout, verify=self.verify_ssl)
        response.raise_for_status()
        return response.text

    def get_drift_summary(self) -> Dict[str, Any]:
        """
        Get drift detection summary.

        Returns:
            Drift detection statistics
        """
        return self._request('GET', '/api/v1/monitoring/drift')

    def get_business_kpis(self) -> Dict[str, Any]:
        """
        Get business KPIs.

        Returns:
            Business metrics dictionary
        """
        return self._request('GET', '/api/v1/monitoring/kpis/business')

    def close(self):
        """Close the session."""
        self.session.close()

    def __enter__(self):
        """Context manager entry."""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit."""
        self.close()


# Convenience function
def predict(text: str, base_url: str = "http://localhost:8000", api_key: Optional[str] = None) -> PredictionResult:
    """
    Quick prediction function.

    Example:
        >>> from kubesentiment_sdk import predict
        >>> result = predict("I love this!")
        >>> print(result.label)
    """
    with KubeSentimentClient(base_url=base_url, api_key=api_key) as client:
        return client.predict(text)
